crates/cranelift/src/compiler/component.rs

```diff
@@ -262,7 +258,163 @@ impl<'a> TrampolineCompiler<'a> {
                 let params = self.builder.func.dfg.block_params(self.block0);
                 (params[2], params[3])
             }
-        };
+        }
+    }
+
+    fn translate_task_return_call(&mut self) {
+        match self.abi {
+            Abi::Wasm => {}
+
+            // These trampolines can only actually be called by Wasm, so
+            // let's assert that here.
+            Abi::Array => {
+                self.builder.ins().trap(TRAP_INTERNAL_ASSERT);
+                return;
+            }
+        }
+
+        let args = self.builder.func.dfg.block_params(self.block0).to_vec();
+        let vmctx = args[0];
+
+        let (values_vec_ptr, values_vec_len) = self.store_wasm_arguments(&args[2..]);
+
+        let (host_sig, index) = host::task_return(self.isa, &mut self.builder.func);
+        let host_fn = self.load_libcall(vmctx, index);
+
+        let params = self.types[self.signature]
+            .unwrap_func()
+            .params()
+            .iter()
+            .map(|&v| {
+                Some(match v {
+                    WasmValType::I32 => FlatType::I32,
+                    WasmValType::I64 => FlatType::I64,
+                    WasmValType::F32 => FlatType::F32,
+                    WasmValType::F64 => FlatType::F64,
+                    _ => return None,
+                })
+            })
+            .collect::<Option<_>>();
+
+        let ty = self.builder.ins().iconst(
+            ir::types::I32,
+            i64::from(
+                params
+                    .and_then(|params| {
+                        self.types
+                            .get_task_return_type(&TypeTaskReturn { params })
+                            .map(|v| v.as_u32())
+                    })
+                    .unwrap_or(u32::MAX),
+            ),
+        );
+
+        let call = self.compiler.call_indirect_host(
+            &mut self.builder,
+            index,
+            host_sig,
+            host_fn,
+            &[vmctx, ty, values_vec_ptr, values_vec_len],
+        );
```

Comment:
You might be able to replace `host::task_return` + `load_libcall` + this with the `call_libcall` helper perhaps?

Reply:
Yeah, I've already done that (and removed other duplication) in a follow-up PR I'm currently finishing up.  Can it wait for that PR?

Reply:
Sounds good 👍